torch/export/exported_program.py

```diff
@@ -338,295 +269,139 @@ def _decompose_and_get_gm_with_new_signature_constants(
-            target = node.target
-            meta = node.meta
-            if node.op == "get_attr":
-                params_buffers_to_node_meta[target] = meta
-
-            # If the call_function uses param as input, we also need to update params' meta
-            # with this call_function node's meta.
-            # This is basically the same flow as torch.fx.traceback.preserve_meta()
-            if node.op == "call_function" and not isinstance(
-                node.target, torch._ops.HigherOrderOperator
-            ):
-                for arg in node._input_nodes:
-                    if arg.op == "get_attr":
-                        for entry in torch.fx.proxy._COPY_META_FIELDS:
-                            if entry in meta:
-                                params_buffers_to_node_meta[arg.target][entry] = meta[
-                                    entry
-                                ]
-
-        with fake_mode:
-            fake_args_unwrapped = pytree.tree_unflatten(fake_args, mod._in_spec)
-            aten_export_artifact = _export_to_aten_ir(
-                mod,
-                # this requires empty kwargs, but not in pytree.flattened format
-                (
-                    *fake_args_unwrapped[0],
-                    *fake_args_unwrapped[1].values(),
-                ),
-                {},
-                fake_params_buffers,
-                constant_attrs,
-                _check_autograd_state=False,
-            )
+    mod.recompile()
 
-        gm = aten_export_artifact.gm
-        new_graph_signature = aten_export_artifact.sig
-
-        for node in gm.graph.nodes:
-            # nn_module_stack
-            if node.op not in ["placeholder", "output"]:
-                for key, (fqn, mod_cls) in node.meta["nn_module_stack"].items():
-                    if isinstance(mod_cls, type):
-                        node.meta["nn_module_stack"][key] = (
-                            fqn,
-                            mod_cls.__module__ + "." + mod_cls.__qualname__,
-                        )
+    # the exported module will store constants & non-persistent buffers such that
+    # retracing treats them as persistent buffers, so we inform the constants lifting pass
+    # and overwrite the new graph signature using the previous program.
+    constant_attrs = ConstantAttrMap()
+    non_persistent_buffers = {
+        spec.target
+        for spec in ep.graph_signature.input_specs
+        if spec.kind == InputKind.BUFFER and not spec.persistent
+    }
+    for name, value in ep.constants.items():
+        if name in non_persistent_buffers:
+            continue
+        # recursive getattr
+        _mod = mod
+        *atoms, attr = name.split(".")
+        for atom in atoms:
+            _mod = getattr(_mod, atom)
+        # remove as buffer, reassign as constant/non-persistent buffer
+        _mod._buffers.pop(attr, None)
+        setattr(_mod, attr, value)
+        constant_attrs.add(value, name)
```

Comment:
Make it a helper function? So the data-dependency is clearer and abstracts away the details.